content/post/2023-06-06-system-dependencies/index.Rmd

```diff
@@ -0,0 +1,259 @@
+
+We now understand more precisely where the magic happens but it still doesn't explain how pak is able to know which precise piece of software to install from the free text `SystemRequirements` field.
+As often when you want to increase your understanding, it is helpful to [read the source](https://blog.r-hub.io/2019/05/14/read-the-source/).
+While browsing pkgdepends source code, we see a call to <https://github.com/r-hub/r-system-requirements>.
+
+This repository contains a set of [rules](https://github.com/rstudio/r-system-requirements/tree/main/rules) as json files which match unformatted software name via regular expressions to the exact libraries for each major operating system.
+Let's walk through an example together:
+
+``` json
+{
+  "patterns": ["\\bnvcc\\b", "\\bcuda\\b"],
+  "dependencies": [
+    {
+      "packages": ["nvidia-cuda-dev"],
+      "constraints": [
+        {
+          "os": "linux",
+          "distribution": "ubuntu"
+        }
+      ]
+    }
+  ]
+}
+```
+
+The regular expression tells that each time a package lists something as `SystemRequirements` with the word "nvcc" or "cuda", the corresponding Ubuntu library to install is `nvidia-cuda-dev`.
+
+## When it's not working out-of-the-box
+
+We are now realizing that this automagical setup we didn't pay so much attention to until now actually requires a very heavy machinery under the hood.
+And it happens, very rarely, that this complex machinery is not able to handle your specific use case.
+But it doesn't mean that you cannot use continuous integration in your package.
+It means that some extra steps might be required to do so.
+Let's review these possible solutions together in order of complexity.
+
+### Fix it for everybody by submitting a pull request
+
+One first option might be that the regular expression used by `r-hub/r-system-requirements` to convert the free text in `SystemRequirements` to a library distributed by your operating system does not recognize what is in `SystemRequirements`.
+
+To identify if this is the case, you need to find the file containing the specific rule for the system dependency of interest in `r-hub/r-system-requirements`, and test the regular expression on the contents of `SystemRequirements`.
+
+If we re-use the cuda example from the previous section and we are wondering why it is not automatically installed for a package specifying "cudaa":
+
+```{r}
+stringr::str_match("cudaa", c("\\bnvcc\\b", "\\bcuda\\b"))
+```
+
+This test confirms that the `SystemRequirements` field contents are not recognized by the regular expression.
+Depending on the case, the best course of action might be to:
+
+-   either edit the contents of `SystemRequirements` so that it's picked up by the regular expression
+-   or submit a pull request to `r-hub/r-system-requirements` if you believe the regular expression is too restrictive and should be updated
```

Comment:
link to an example PR?

Reply:
@gaborcsardi, could you confirm that `r-hub/r-system-requirements` is really the repo where PR should be submitted? Or does it make more sense to submit it to the upstream `rstudio/r-system-requirements`?

Reply:
It should be submitted upstream at `rstudio/r-system-requirements`.

Reply:
@Bisaloo In fact, I wouldn't even mention `r-hub/r-system-requirements`, only the upstream repo.

Reply:
Great, thanks for your fast answer! This also means that changes will take time to propagate even when merged upstream? Or do you automatically regularly update the fork?

Reply:
Changes do take time, nothing uses that repo "live", so PPM changes take time, and pak changes take time as well, and yes, the latter are manual. 
This is to avoid having an upstream breaking change, e.g. a schema change, breaking pak.